lib/job.js

```diff
@@ -198,22 +198,29 @@ Job.prototype.promote = function(){
     });
 };
 
-Job.prototype.retry = function(){
-  var key = this.queue.toKey('wait');
-  var failed = this.queue.toKey('failed');
-  var channel = this.queue.toKey('jobs');
-  var multi = this.queue.multi();
+Job.prototype.retry = function() {
   var _this = this;
-
-  multi.srem(failed, this.jobId);
-  // if queue is LIFO use rpushAsync
-  multi[(this.opts.lifo ? 'r' : 'l') + 'push'](key, this.jobId);
-  multi.publish(channel, this.jobId);
-
-  return multi.execAsync().then(function(){
-    _this.queue.emit('waiting', _this);
-    return _this;
+  var push = (this.opts.lifo ? 'R' : 'L') + 'PUSH';
+  var script = [
+    'if (redis.call("SREM", KEYS[2], ARGV[1]) == 1) or (redis.call("SREM", KEYS[3], ARGV[1]) == 1) then',
+    ' redis.call("' + push + '", KEYS[1], ARGV[1])',
+    ' redis.call("PUBLISH", KEYS[4], ARGV[1])',
+    ' return 0',
+    'end',
+    'return -1'
+  ].join('\n');
+  var keys = ['wait', 'failed', 'completed', 'jobs'].map(function(name) {
```

Comment:
Shouldn't the job be added moved the `delay` queue? By skipping it directly to `wait` it does not honor the delay option (if specified).